File: array/a9.py
# Topic: Minimise the maximum difference between heights 

from typing import List, Any


class Array:
    def __init__(self, elements: List[Any]) -> None:
        self.elements = elements
    
    def get_min_diff(self, n: int, k: int) -> int:
        self.elements.sort()
        ans   = self.elements[n-1] - self.elements[0]
        small = self.elements[0] + k
        big   = self.elements[n-1] - k
        if small > big:
            small, big = big, small

        for element in self.elements[1:]:
            add_k = element + k
            sub_k = element - k

            if (sub_k >= small or add_k >= big):
                continue
            if (big - small <= add_k - sub_k):
                small = sub_k
            else:
                big   = add_k

        return min((ans, big - small))

array = Array([1, 5, 8, 10])

print("Minimize the heights: ", array.get_min_diff(
    len(array.elements), 2
))


def test_a9():
    array = Array([3, 9, 12, 16, 20])
    assert array.get_min_diff(5, 3) == 11

File: array/a9.rs
// Topic: Minimise the maximum difference between heights 

use std::cmp::min;

struct Array { elements: Vec<i64> }

impl Array {
    fn get_min_diff(&mut self, k: i64) -> i64 {
        self.elements.sort();
        let len_elems = self.elements.len();
        let ans = self.elements[len_elems - 1] - self.elements[0];
        let mut small = self.elements[0] + k;
        let mut big = self.elements[len_elems - 1] - k;
        if small > big {
            small = big;
            big =small;
        }

        for element in &self.elements[1..] {
            let add = *element + k;
            let sub = *element - k;

            if sub >= small || add >= big {
                continue;
            }

            if big - small <= add - sub {
                small = sub
            } else {
                big = add
            }
        }
        return min(ans, big - small);
    }
}

fn main() {
    let mut array = Array{ elements: vec![3, 9, 12, 16, 20] };
    println!("Min heights: {}", array.get_min_diff(3))
}
